server/src/main/java/org/opensearch/index/engine/EngineConfig.java

```diff
@@ -133,15 +133,15 @@ public Supplier<RetentionLeases> retentionLeasesSupplier() {
             case "lz4":
             case "best_compression":
             case "zlib":
-            case "zstd":
-            case "zstd_no_dict":
             case "lucene_default":
                 return s;
             default:
+                if (("zstd".equals(s) || "zstd_no_dict".equals(s)) && Codec.availableCodecs().contains("Lucene95CustomCodec")) {
```

Comment:
It doesn't look right: if plugin is not installed, we should fail. However, when plugin is installed, "zstd" and "zstd_no_dict" should appear in `Codec.availableCodecs()`

Reply:
There was a miss in how it was implemented earlier, ```zstd``` and ```zstd_no_dict``` are actually two different modes of the same codec ```Lucene95CustomCodec```, but it was made as top level codecs earlier (both in sandbox and when it moved to the core). When the custom-codecs plugin is installed ```Lucene95CustomCodec``` will be part of ```Codec.availableCodecs()``` but the modes are internal to the plugin and hence not visible to the core. To work around that (for backward compatibility) this (hacky) change is required.

Reply:
@reta an additional integ test has been added to ensure that if the plugin is not installed, a user will not be able create indices with ZStd codecs.

Reply:
> There was a miss in how it was implemented earlier, zstd and zstd_no_dict are actually two different modes of the same codec Lucene95CustomCodec, but it was made as top level codecs earlier (both in sandbox and when it moved to the core).

@sarthakaggarwal97 you are missing the codecs services https://github.com/opensearch-project/OpenSearch/pull/3577/files#diff-5026012c796bed585ea50d926ee4155d8283251009a9d70e2897f4e43667200e, the core one should be cleaned out  https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/resources/META-INF/services/org.apache.lucene.codecs.Codec, the core should not refer to "zstd" (or any custom codec from plugin in general)

Reply:
@sarthakaggarwal97 do you still have issues with Codec.availableCodecs()? we should not refer to "zstd" (or any custom codec from plugin in general)

Reply:
@reta yeah I'm afraid there might not be any other way. With Lucene's Codec, Lucene95CustomCodec is being registered currently. So, when we do the validation in settings, we are don't see zstd or zstd_no_dict codec over there, instead we see Lucene95CustomCodec.
Similar thing happens in case of BEST_SPEED and BEST_COMPRESSION, we see Lucene95Codec registered with Lucene.
Also, zstd_no_dict as a name, cannot be registered with the Lucene due its SPI rules (only alpha-numberic characters allowed).
When we moved the codecs to core, we added them to the switch case, which made it possible to give the user a friendly name with underscores.

Reply:
Thanks @sarthakaggarwal97 , let's keep it this way for now, I am looking into it

Reply:
Yeah, this was missed in the original PR (another reason it should've been feature flagged).  `ZSTD` and `ZSTD_NO_DICT` should be registered through `NamedSPI` as two different codecs

Reply:
Just remove the `Lucene95CustomCodec#Mode` enum altogether and register this as a separate "named" codec with the same name.

Reply:
> `ZSTD` and `ZSTD_NO_DICT` should be registered through `NamedSPI` as two different codecs

My understanding it was using NamedSPI (META-INF/services/org.apache.lucene.codecs.Codec), or you mean something else?

Reply:
@nknize I think we won't be able to register ```ZSTD_NO_DICT``` with Lucene's NamedSPILoader due to [this](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/util/NamedSPILoader.java#L82).
Also, we would be breaking user experience if we go ahead with ```ZSTDNODICT``` or something similar

Reply:
also, we will have to move the ZStd from the ```opensearch-compress``` library right as a part of moving ZStd out of core?

Reply:
> My understanding it was using NamedSPI (META-INF/services/org.apache.lucene.codecs.Codec), or you mean something else? 

No. by "named" codec.. I mean through `NamedSPI` in lucene per my [comment](https://github.com/opensearch-project/OpenSearch/pull/9658/files#r1313246952) right above.

> I think we won't be able to register ZSTD_NO_DICT with Lucene's NamedSPILoader...

Oye, that's right. Just write a simple utility method to convert it to a lucene compliant name then...  `private String convertToLuceneName(final String opensearchCodecName) {`

Reply:
> also, we will have to move the ZStd from the `opensearch-compress` library right as a part of moving ZStd out of core?

No... just remove the dependency in [`server/build.gradle`](https://github.com/opensearch-project/OpenSearch/blob/2fb46940ddeaadf186ba60a4f6a9fc7fe77259f0/server/build.gradle#L104). My [compressor library PR](https://github.com/opensearch-project/OpenSearch/pull/9262) already decoupled the compressors for `BlobStoreRepository` from `server` by using SPI. Having `zstd-jni` in the compressor library is still optional. 

Although it is dangerous because at any time someone could take a dependency on that library and have the ZStdCompressor jni risks. This is another reason why I [suggested moving `custom-codecs` plugin](https://github.com/opensearch-project/OpenSearch/issues/9422#issuecomment-1702972544) to either `common-util` or a new `custom-compression` library.